sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_kernels.asciidoc

```diff
@@ -0,0 +1,974 @@
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 8 specification.
+All references below to the "core SYCL specification" or to section numbers in
+the SYCL specification refer to that revision.
+
+This extension also depends on the following other SYCL extensions:
+
+* link:../proposed/sycl_ext_oneapi_free_function_queries.asciidoc[
+  sycl_ext_oneapi_free_function_queries]
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.
+Interfaces defined in this specification may not be implemented yet or may be
+in a preliminary state.
+The specification itself may also change in incompatible ways before it is
+finalized.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+
+== Overview
+
+This extension introduces a new way to define a kernel as a simple C++
+function, where the kernel arguments are parameters to the function.
+This is different from standard SYCL kernels, where the kernel arguments are
+either captures of a lambda expression or member variables of a callable
+object.
+
+The primary motivation for this extension is the
+link:../proposed/sycl_ext_oneapi_kernel_compiler.asciidoc[
+sycl_ext_oneapi_kernel_compiler], which allows online compilation of a kernel
+from source code.
+These kernels must have a clean separation between host and device code, and
+they must have a defined order to their arguments because the application sets
+the argument values by their "index" via `handler::set_arg(index, value)`.
+Because variables captured by lambda expressions have no defined order, we need
+some other way to define kernel arguments for use with the
+sycl_ext_oneapi_kernel_compiler extension, and the
+sycl_ext_oneapi_free_function_kernels extension provides that mechanism.
+The "free function kernel" feature is a separate extension, though, because it
+can also be used independently from sycl_ext_oneapi_kernel_compiler.
+For example, some users may find it more familiar to define kernels as plain
+functions or they might have other reasons to prefer a clean separation between
+host and device code.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.
+An implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_FREE_FUNCTION_KERNELS`
+to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Defining a free function kernel
+
+A free function kernel is a normal C++ function definition, where the function
```

Comment:
This comment made me wonder whether the term "free function" is really want we want to use here.  I'm not sure there is a strict definition for this term.  As far as I can tell, it's not a term from the C++ specification.

This extension allows the function to be defined either at namespace scope or at class scope as a static member function.  (This is stated in the first bullet item below.)

I *think* people usually use the term "free function" to refer to a function that is not a member function.  In that case, "C++ free function" is not really the right term because we also allow the function to be a static member function.  Therefore, I think we should not change the spec to say the function must be a "C++ free function".

This also makes me wonder whether we should rename the extension to something like "plain function kernels".  I don't know of a formal concise term that means "a function declared at namespace scope or a static member function".  We could also just keep the current extension name.  I think the text of the extension clearly states which functions are allowed.

Reply:
Yeah I see what you mean, "free function" might not be the best term for this. I think "free function" is the colloquial term that's used, but the C++ specification seems to use the term "non-member function".

So yeah this extension is supporting non-member functions and static member functions (as well as non-member function templates and static member function templates), I don't think there is a collective term for this.

I think "plain function kernels" is a good alternative, but it maybe excludes templates, maybe it's better to just to name them "function kernels"?

Reply:
We had a phone conversation about this and decided that the term "function kernel" is also confusing because the core SYCL spec uses the term "kernel function" to refer to the kernels that are defined as a lambda expression or named kernel object.  Therefore, we decided to just continue with the current name "free function kernel".